Realm/ObjectStore/impl/realm_coordinator.cpp

```diff
@@ -0,0 +1,453 @@
+
+    clean_up_dead_queries();
+
+    if (m_queries.empty() && m_new_queries.empty()) {
+        return;
+    }
+
+    if (!m_async_error) {
+        open_helper_shared_group();
+    }
+
+    if (m_async_error) {
+        move_new_queries_to_main();
+        return;
+    }
+
+    advance_helper_shared_group_to_latest();
+
+    // Make a copy of the queries vector so that we can release the lock while
+    // we run the queries
+    auto queries_to_run = m_queries;
+    lock.unlock();
+
+    for (auto& query : queries_to_run) {
+        query->run();
+    }
+
+    // Reacquire the lock while updating the fields that are actually read on
+    // other threads
+    {
+        lock.lock();
+        for (auto& query : queries_to_run) {
+            query->prepare_handover();
+        }
+    }
+
+    clean_up_dead_queries();
+}
+
+void RealmCoordinator::open_helper_shared_group()
+{
+    if (!m_query_sg) {
+        try {
+            std::unique_ptr<Group> read_only_group;
+            Realm::open_with_config(m_config, m_query_history, m_query_sg, read_only_group);
+            REALM_ASSERT(!read_only_group);
+            m_query_sg->begin_read();
+        }
+        catch (...) {
+            // Store the error to be passed to the async queries
+            m_async_error = std::current_exception();
+            m_query_sg = nullptr;
+            m_query_history = nullptr;
+        }
+    }
+    else if (m_queries.empty()) {
+        m_query_sg->begin_read();
+    }
+}
+
+void RealmCoordinator::move_new_queries_to_main()
+{
+    m_queries.reserve(m_queries.size() + m_new_queries.size());
+    std::move(m_new_queries.begin(), m_new_queries.end(), std::back_inserter(m_queries));
+    m_new_queries.clear();
+}
+
+void RealmCoordinator::advance_helper_shared_group_to_latest()
+{
+    if (m_new_queries.empty()) {
+        LangBindHelper::advance_read(*m_query_sg, *m_query_history);
+        return;
+    }
+
+    // Sort newly added queries by their source version so that we can pull them
+    // all forward to the latest version in a single pass over the transaction log
+    std::sort(m_new_queries.begin(), m_new_queries.end(), [](auto const& lft, auto const& rgt) {
+        return lft->version() < rgt->version();
+    });
+
+    // Import all newly added queries to our helper SG
+    for (auto& query : m_new_queries) {
+        LangBindHelper::advance_read(*m_advancer_sg, *m_advancer_history, query->version());
+        query->attach_to(*m_advancer_sg);
+    }
+
+    // Advance both SGs to the newest version
+    LangBindHelper::advance_read(*m_advancer_sg, *m_advancer_history);
+    LangBindHelper::advance_read(*m_query_sg, *m_query_history,
+                                 m_advancer_sg->get_version_of_current_transaction());
+
+    // Transfer all new queries over to the main SG
+    for (auto& query : m_new_queries) {
+        query->detatch();
+        query->attach_to(*m_query_sg);
+    }
+
+    if (!m_new_queries.empty()) {
```

Comment:
Is it possible for `m_new_queries` to be empty here even though we already checked for that case in the early-return at the start of this function?